KAFKA-9550 Copying log segments to tiered storage in RemoteLogManager

satishd opened the pull request:
Added functionality to copy log segments, indexes to the target remote storage for each topic partition enabled with tiered storage. This involves creating scheduled tasks for all leader partition replicas to copy their log segments in sequence to tiered storage.

Other contributors: @kamalcph , @showuon 
@kamalcph Minor improvements in RemoteLogManager and a few tests in RemoteLogManagerTest
@showuon contributed to adding a few more unit tests to the existing RemoteLogManagerTest. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

showuon commented:
Thanks for the PR. I'll have an early review this weekend. :)

satishd committed:
KAFKA-9550 Copying log segments to tiered storage in RemoteLogManager

Added functionality to copy log segments, indexes to the target remote storage for each topic partition enabled with tiered storage. This involves creating scheduled tasks for all leader partition replicas to copy their log segments in sequence to tiered storage.t push -f origin

showuon commented on storage/src/main/java/org/apache/kafka/storage/internals/epoch/LeaderEpochFileCache.java:
```diff
@@ -359,6 +361,27 @@ public OptionalInt epochForOffset(long offset) {
+        }
+    }
+
+    public OptionalInt findPreviousEpoch(int epoch) {
```
Any reason we don't use the existing `previousEpoch` method?

satishd replied:
Good catch! This method is already available as `previousEpoch(int epoch)`.

satishd committed:
Addressed review comments
 - Using the existing previousEpoch(int epoch) in LeaderEpochFileCache
 - Minor cleanup of removing duplicated code

showuon commented on core/src/main/scala/kafka/log/UnifiedLog.scala:
```diff
@@ -520,6 +522,11 @@ class UnifiedLog(@volatile var logStartOffset: Long,
   }
+  def updateHighestOffsetInRemoteStorage(offset: Long): Unit = {
+    if (!remoteLogEnabled())
+      warn(s"Received update for highest offset with remote index as: $offset, the existing value: $highestOffsetInRemoteStorage")
```
It's unclear what we tried to warn? Maybe something like this:
`warn(s"Unable to update the highest offset with remote index as: $offset, the existing value: $highestOffsetInRemoteStorage since remote storage is not enabled.")`

showuon commented on core/src/main/scala/kafka/log/remote/RemoteLogManager.scala:
```diff
@@ -17,39 +17,107 @@
+  private val scheduledThreadPool: ScheduledThreadPoolExecutor = {
+    val threadPool = new ScheduledThreadPoolExecutor(poolSize)
+    threadPool.setRemoveOnCancelPolicy(true)
+    threadPool.setExecuteExistingDelayedTasksAfterShutdownPolicy(false)
```
should we set `setContinueExistingPeriodicTasksAfterShutdownPolicy(false);` ?

showuon commented on core/src/main/scala/kafka/log/remote/RemoteLogManager.scala:
```diff
@@ -17,39 +17,107 @@
+  def shutdown(): Boolean = {
+    info("Shutting down scheduled thread pool")
+    scheduledThreadPool.shutdownNow()
+    //waits for 2 mins to terminate the current tasks
```
This comment is not needed.

showuon commented on core/src/main/scala/kafka/log/remote/RemoteLogManager.scala:
```diff
@@ -17,39 +17,107 @@
  *
  * @param rlmConfig Configuration required for remote logging subsystem(tiered storage) at the broker level.
  * @param brokerId  id of the current broker.
  * @param logDir    directory of Kafka log segments.
```
Please remember to update the javadoc

showuon commented on core/src/main/scala/kafka/log/remote/RemoteLogManager.scala:
```diff
@@ -272,15 +310,326 @@ class RemoteLogManager(rlmConfig: RemoteLogManagerConfig,
     None
   }
 
+  trait CancellableRunnable extends Runnable {
```
Why should we declare `CancellableRunnable` twice in this class?

showuon reviewed (commented):
Left some comments. Will continue to review the remaining changes later.

satishd committed:
Minor refactoring in RLM

showuon commented on core/src/main/scala/kafka/log/remote/RemoteLogManager.scala:
```diff
@@ -143,24 +216,44 @@ class RemoteLogManager(rlmConfig: RemoteLogManagerConfig,
+    def cacheTopicPartitionIds(topicIdPartition: TopicIdPartition): Unit = {
+      val previousTopicId = topicPartitionIds.put(topicIdPartition.topicPartition(), topicIdPartition.topicId())
+      if (previousTopicId != null && previousTopicId != topicIdPartition.topicId()) {
+        warn(s"Previous cached topic id $previousTopicId for ${topicIdPartition.topicPartition()} does " +
```
This warning message seems not necessary IMO. Users cannot do anything when getting this info, right?

satishd replied:
Changed it to info level, which may be useful for debugging.

showuon commented on core/src/main/scala/kafka/log/remote/RemoteLogManager.scala:
```diff
@@ -272,15 +364,270 @@ class RemoteLogManager(rlmConfig: RemoteLogManagerConfig,
+          if (lso < 0) {
+            warn(s"lastStableOffset for partition $tpId is $lso, which should not be negative.")
+          } else if (lso > 0 && copiedOffset < lso) {
+            // Copy segments only till the min of high-watermark or stable-offset as remote storage should contain
+            // only committed/acked messages
+            val toOffset = lso
```
Where do we set the offset with `min of high-watermark or stable-offset`?

satishd replied:
LastStableOffset already represents that, will update the javadoc comment.

showuon commented on core/src/main/scala/kafka/log/remote/RemoteLogManager.scala:
```diff
@@ -272,15 +364,270 @@ class RemoteLogManager(rlmConfig: RemoteLogManagerConfig,
+                val producerIdSnapshotFile: File = log.producerStateManager.fetchSnapshot(nextOffset).orElse(null)
+
+                val epochEntries = getLeaderEpochCheckpoint(log, segment.baseOffset, nextOffset).read()
+                val segmentLeaderEpochs: util.Map[Integer, lang.Long] = new util.HashMap[Integer, lang.Long](epochEntries.size())
```
Should we use `lang.Integer` here?

satishd replied:
The existing `Integer` represents that, adding `lang` is redundant. We had to add `lang` to Long as it conflicts with scala's `Long`.

showuon replied:
Ah, you're right!

showuon commented on core/src/main/scala/kafka/log/remote/RemoteLogManager.scala:
```diff
@@ -272,15 +364,270 @@ class RemoteLogManager(rlmConfig: RemoteLogManagerConfig,
+                  segmentLeaderEpochs.put(Integer.valueOf(entry.epoch), lang.Long.valueOf(entry.startOffset))
+                )
+
+                val remoteLogSegmentMetadata = new RemoteLogSegmentMetadata(id, segment.baseOffset, endOffset,
```
could we rename this variable to a meaningful name like `copySegmentStartedMetadata`?

showuon reviewed (commented):
Left some more comments.

showuon commented on core/src/main/scala/kafka/log/remote/RemoteLogManager.scala:
```diff
@@ -272,15 +364,270 @@ class RemoteLogManager(rlmConfig: RemoteLogManagerConfig,
+      val future = rlmScheduledThreadPool.scheduleWithFixedDelay(task, 0, delayInMs, TimeUnit.MILLISECONDS)
+      RLMTaskWithFuture(task, future)
+    })
+    convertToLeaderOrFollower(rlmTaskWithFuture.rlmTask)
```
Will there be race condition when the topic partition is new inserted, so it'll set `convertToLeaderOrFollower` once before scheduling. And then the scheduled thread could be run immediately, later, run this line `convertToLeaderOrFollower(rlmTaskWithFuture.rlmTask)`, so it reset the leader epoch if it got updated. 

Could we add a null condition check for `rlmTaskWithFuture` here? that is:
```
if (rlmTaskWithFuture != null) {
  // not new created task, should convert it here
 convertToLeaderOrFollower(rlmTaskWithFuture.rlmTask)
}
```

WDYT?

satishd replied:
There is no race condition here. `rlmTaskWithFuture` will never be null as it will be created if it does not exist in `leaderOrFollowerTasks` map. It is expected that leader or follower switches can happen resulting leader epoch updates and `RLMTask` takes care of handling those updates.

satishd committed:
Addressed review comments

junrao commented on core/src/main/scala/kafka/log/remote/RemoteLogManager.scala:
```diff
@@ -17,39 +17,97 @@
  * @param rlmConfig Configuration required for remote logging subsystem(tiered storage) at the broker level.
  * @param brokerId  id of the current broker.
  * @param logDir    directory of Kafka log segments.
+ * @param time      Time instance.
```
Since this is a relatively new class and we are in the process of migrating scala classes to java, would it be better to write this class in java but still keep it in the core module? This way, we don't need to write the new code twice, once in scala and another in java.

satishd replied:
This was an existing code in Scala in my other branch with a few changes. But your suggestion is better as it avoids the review again during Java migration later. 
I did the Java migration as you suggested and pushed it to the PR.

junrao reviewed (commented):
@satishd : Thanks for the PR. Left a quick comment.

satishd committed:
Addressed review comments
Migrated RemoteLogManager to java in core module

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,749 @@
+        }
+    }
+
+    RemoteStorageManager createRemoteStorageManager() throws Exception {
```
This method doesn't seem to throw Exception? Also, could this be private?

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,749 @@
+    private void cacheTopicPartitionIds(TopicIdPartition topicIdPartition) {
+        Uuid previousTopicId = topicPartitionIds.put(topicIdPartition.topicPartition(), topicIdPartition.topicId());
+        if (previousTopicId != null && previousTopicId != topicIdPartition.topicId()) {
+            LOGGER.info("Previous cached topic id {} for {} does not match update topic id {}",
```
update => the updated ?

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,749 @@
+                    LOGGER.info("Created a new task: {} and getting scheduled", task);
+                    ScheduledFuture future = rlmScheduledThreadPool.scheduleWithFixedDelay(task, 0, delayInMs, TimeUnit.MILLISECONDS);
+                    return new RLMTaskWithFuture(task, future);
+
```
unnecessary extra line.

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,749 @@
+        public RLMTask(TopicIdPartition topicIdPartition) {
+            this.topicIdPartition = topicIdPartition;
+            LogContext logContext = new LogContext("[RemoteLogManager=" + brokerId + " partition=" + topicIdPartition + "] ");
+            logger = logContext.logger(LeaderEpochFileCache.class);
```
Wrong class?

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,749 @@
+            return leaderEpoch >= 0;
+        }
+
+        // The readOffset is None initially for a new leader RLMTask, and needs to be fetched inside the task's run() method.
```
None => empty?

Also, there is no readOffset. Should copiedOffsetOption be readOffset?

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,749 @@
+
+        public void convertToLeader(int leaderEpochVal) {
+            if (leaderEpochVal < 0) {
+                throw new KafkaException("leaderEpoch value for topic partition $tpId can not be negative");
```
There is no tpId.

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,749 @@
+                // of a segment with that epoch copied into remote storage. If it can not find an entry then it checks for the
+                // previous leader epoch till it finds an entry, If there are no entries till the earliest leader epoch in leader
+                // epoch cache then it starts copying the segments from the earliest epoch entry’s offset.
+                copiedOffsetOption = OptionalLong.of(findHighestRemoteOffset(topicIdPartition));
```
`findHighestRemoteOffset` could return -1. Should we handle that?

satishd replied:
This is already handled while finding the `fromOffset` like below.
```
long fromOffset = Math.max(copiedOffset + 1, log.logStartOffset());
```

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,749 @@
+
+                UnifiedLog log = maybeLog.get();
+
+                // LSO indicates the offset below are ready to be consumed(high-watermark or committed)
```
space before (

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,749 @@
+                    // 1) log-start-offset gets incremented via delete-records API (or)
+                    // 2) enabling the remote log for the first time
+                    long fromOffset = Math.max(copiedOffset + 1, log.logStartOffset());
+                    ArrayList<LogSegment> sortedSegments = new ArrayList<>(JavaConverters.asJavaCollection(log.logSegments(fromOffset, toOffset)));
```
JavaConverters is deprecated. Could we use CollectionConverters instead?

satishd replied:
`CollectionConverters` is only available in Scala-2.13. We used `JavaConverters` to support both 2.12 and 2.13.

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,749 @@
+                    ArrayList<LogSegment> sortedSegments = new ArrayList<>(JavaConverters.asJavaCollection(log.logSegments(fromOffset, toOffset)));
+                    sortedSegments.sort(Comparator.comparingLong(LogSegment::baseOffset));
+                    List<Long> sortedBaseOffsets = sortedSegments.stream().map(x -> x.baseOffset()).collect(Collectors.toList());
+                    int activeSegIndex = Collections.binarySearch(sortedBaseOffsets, activeSegBaseOffset);
```
This seems a bit complicated. Could we just set toOffset to the min of lso and activeSegBaseOffset?

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,749 @@
+            logger.info("Copying {} to remote storage.", logFileName);
+            RemoteLogSegmentId id = new RemoteLogSegmentId(topicIdPartition, Uuid.randomUuid());
+
+            long nextOffset = segment.readNextOffset();
```
`readNextOffset() ` could be a bit expensive. Could we get nextOffset from the base offset of the next segment?

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,749 @@
+     * @return the truncated leader epoch checkpoint
+     */
+
+    InMemoryLeaderEpochCheckpoint getLeaderEpochCheckpoint(UnifiedLog log, long startOffset, long endOffset) {
```
Could this be private?

satishd replied:
This is needed for tests, which are added now.

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,749 @@
+            remoteLogMetadataManager.addRemoteLogSegmentMetadata(copySegmentStartedRlsm).get();
+
+            ByteBuffer leaderEpochsIndex = getLeaderEpochCheckpoint(log, -1, nextOffset).readAsByteBuffer();
+            LogSegmentData segmentData = new LogSegmentData(logFile.toPath(), toPathIfExists(segment.lazyOffsetIndex().get().file()),
```
Instead of passing in null `Path`, would it be better to pass in `Optional`?

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,749 @@
+
+            long nextOffset = segment.readNextOffset();
+            long endOffset = nextOffset - 1;
+            File producerIdSnapshotFile = log.producerStateManager().fetchSnapshot(nextOffset).orElse(null);
```
producerIdSnapshotFile => producerStateSnapshotFile ?

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,749 @@
+                    producerIdSnapshotFile.toPath(), leaderEpochsIndex);
+            remoteLogStorageManager.copyLogSegmentData(copySegmentStartedRlsm, segmentData);
+
+            RemoteLogSegmentMetadataUpdate copySegmentFinishedRlsm = new RemoteLogSegmentMetadataUpdate(id, time.milliseconds(),
```
copySegmentFinishedRlsm => finishedSegmentRlsm ?

satishd replied:
Does not `copySegmentFinishedRlsm` represent what it is meant for instead of `finishedSegmentRlsm `?

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,749 @@
+                RecordBatch batch = remoteLogInputStream.nextBatch();
+                if (batch == null) break;
+                if (batch.maxTimestamp() >= timestamp && batch.lastOffset() >= startingOffset) {
+                    for (Record record : batch) {
```
remove space before : ?

satishd replied:
I think the default code format is about having space before and after ':'. I see the same in all other places.

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,749 @@
+        }
+    }
+
+    public long findHighestRemoteOffset(TopicIdPartition topicIdPartition) throws RemoteStorageException {
```
Could this be private?

satishd replied:
This is needed for tests which are added now.

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,749 @@
+            return threadPool;
+        }
+
+        public void resizePool(int size) {
```
This seems unused?

satishd replied:
This is used for dynamic config changes that we planned to add later, removed it for now.

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,749 @@
+            scheduledThreadPool.setCorePoolSize(size);
+        }
+
+        public int poolSize() {
```
This seems unused?

satishd replied:
This is used for dynamic config changes that we planned to add later, removed it for now.

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,749 @@
+        }
+
+        public String toString() {
+            return this.getClass().toString() + "[" + topicIdPartition + "]";
```
Should we include the segment base offset too?

satishd replied:
base offset changes in RLMTask, and it is logged now. 
We plan to add a metric about tier offset lag. This is already implemented internally. We will raise a PR in future.

junrao commented on server-common/src/main/java/org/apache/kafka/server/common/CheckpointFile.java:
```diff
@@ -77,20 +77,8 @@ public void write(Collection<T> entries) throws IOException {
-                    writer.newLine();
-                }
-
+                CheckpointWriteBuffer<T> checkpointWriteBuffer = new CheckpointWriteBuffer<T>(writer, version, formatter);
```
new CheckpointWriteBuffer<T> => new CheckpointWriteBuffer<> ?

junrao commented on storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java:
```diff
@@ -605,6 +609,8 @@ public Optional<SnapshotFile> removeAndMarkSnapshotForDeletion(long snapshotOffs
         return Optional.empty();
     }
 
+
```
unnecessary new lines.

junrao commented on core/src/main/scala/kafka/log/UnifiedLog.scala:
```diff
@@ -520,6 +522,11 @@ class UnifiedLog(@volatile var logStartOffset: Long,
   }
+  def updateHighestOffsetInRemoteStorage(offset: Long): Unit = {
+    if (!remoteLogEnabled())
+      warn(s"Unable to update the highest offset with remote index as: $offset, the existing value: $highestOffsetInRemoteStorage since remote storage is not enabled.")
```
Could we make it more readable like the following?

Unable to update the highest offset in remote storage with offset $offset since remote storage is not enabled. The existing highest offset is $highestOffsetInRemoteStorage.

junrao commented on core/src/main/scala/kafka/log/UnifiedLog.scala:
```diff
@@ -1231,10 +1238,10 @@ class UnifiedLog(@volatile var logStartOffset: Long,
 
           remoteLogManager.get.findOffsetByTimestamp(topicPartition, targetTimestamp, logStartOffset, leaderEpochCache.get)
-        } else None
+        } else Optional.empty()
```
We should return None, right?

satishd replied:
Earlier `remoteLogManager#findOffsetByTimestamp` was returning Scala `Option` but it returns `Optional` now after it is migrated to Java.

junrao reviewed (commented):
@satishd : Thanks for the PR. A few comments below.

junrao commented:
@satishd : Do you plan to add any new tests?

showuon committed:
Refactored RemoteLogManagerTest from Scala to Java and added a few tests.

satishd committed:
Addressed review comments

satishd committed:
Minor cleanup in RemoteLogManagerTest

showuon committed:
Added copyLogSegment test

satishd committed:
- Minor cleanup with addressing review comments
- Use JavaConverters instead of CollectionConverters as it is not available in Scala-2.12

satishd commented:
Thanks @junrao for your review comments. Some of those comments are addressed inline or with the updated commits, will address the remaining few comments soon.

> @satishd : Do you plan to add any new tests?
As I mentioned in the earlier PR description, that was a 'todo' to be done in this PR. They are added now with the latest commits.

showuon commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,736 @@
+     * @param logDir    directory of Kafka log segments.
+     * @param time      Time instance.
+     * @param fetchLog  function to get UnifiedLog instance for a given topic.
+     * @throws Exception
```
This is not necessary if we don't specify which exception will be thrown and when.

showuon commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,736 @@
+     * @param partitionsBecomeFollower partitions that have become followers on this broker.
+     * @param topicIds                 topic name to topic id mappings.
+     */
+    public void onLeadershipChange(Set<Partition> partitionsBecomeLeader,
```
I didn't see any place we invoke this method. I think `ReplicaManager#becomeLeaderOrFollower` should be responsible to invoke this method, right?

showuon commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,736 @@
+
+    }
+
+    static class InMemoryLeaderEpochCheckpoint implements LeaderEpochCheckpoint {
```
This is not necessary, and use `org.apache.kafka.storage.internals.checkpoint.InMemoryLeaderEpochCheckpoint` instead.

showuon reviewed (commented):
Left some comments. Thanks for the PR!

satishd committed:
Addressed review comments

showuon reviewed (approved):
LGTM!

showuon commented:
Failed tests are unrelated:
```
    Build / JDK 17 and Scala 2.13 / kafka.admin.TopicCommandIntegrationTest.testDescribeAndListTopicsWithoutInternalTopics(String).quorum=zk
    Build / JDK 17 and Scala 2.13 / kafka.api.TransactionsTest.testFailureToFenceEpoch(String).quorum=kraft
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.integration.OffsetsApiIntegrationTest.testGetSinkConnectorOffsetsDifferentKafkaClusterTargeted
    Build / JDK 11 and Scala 2.13 / kafka.security.authorizer.AuthorizerTest.testSuperUserHasAccess(String).quorum=kraft
    Build / JDK 11 and Scala 2.13 / kafka.server.DynamicBrokerReconfigurationTest.testUncleanLeaderElectionEnable()
    Build / JDK 8 and Scala 2.12 / kafka.security.authorizer.AuthorizerTest.testNoAclFoundOverride(String).quorum=kraft
    Build / JDK 8 and Scala 2.12 / kafka.security.authorizer.AuthorizerTest.testSuperUserWithCustomPrincipalHasAccess(String).quorum=kraft
    Build / JDK 8 and Scala 2.12 / kafka.security.authorizer.AuthorizerTest.testDeleteAclOnWildcardResource(String).quorum=kraft
```

Since this PR blocks other following PRs, I'm going to merge it soon. We can have a follow-up PR to address any new comments if any. Thank you.
cc @junrao

Hangleton commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,719 @@
+
+        remoteLogStorageManager = createRemoteStorageManager();
+        remoteLogMetadataManager = createRemoteLogMetadataManager();
+        indexCache = new RemoteIndexCache(1024, remoteLogStorageManager, logDir);
```
Why is the size of the remote index cache 1,024? It may not be sufficient for all use cases. Maybe, this could be configurable via a property.

Hangleton commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,719 @@
+    private final ConcurrentHashMap<TopicIdPartition, RLMTaskWithFuture> leaderOrFollowerTasks = new ConcurrentHashMap<>();
+
+    // topic ids that are received on leadership changes, this map is cleared on stop partitions
+    private final ConcurrentMap<TopicPartition, Uuid> topicPartitionIds = new ConcurrentHashMap<>();
```
I wonder if we could rely on the topic-id mapping in the `MetadataCache` instead of maintaining the mapping locally. Is there any benefit in doing that here, versus using the `MetadataCache` as a source of truth?

Hangleton commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,719 @@
+     * @return the timestamp and offset of the first message that meets the requirements. None will be returned if there
+     * is no such message.
+     */
+    public Optional<FileRecords.TimestampAndOffset> findOffsetByTimestamp(TopicPartition tp,
```
Maybe all the methods/contracts of the `RemoteLogManager` could be based on `TopicIdPartition`. It would then be the responsibility of the caller to perform the resolution of the topic ids if the Kafka API for which the invocation of the RLM is made does not support it.

Hangleton commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,719 @@
+    }
+
+    RemoteLogMetadataManager createRemoteLogMetadataManager() {
+        return AccessController.doPrivileged(new PrivilegedAction<RemoteLogMetadataManager>() {
```
(nit) `AccessController` has been tagged deprecated and for removal since JDK 17.

Hangleton commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,719 @@
+            if (leaderEpochVal < 0) {
+                throw new KafkaException("leaderEpoch value for topic partition " + topicIdPartition + " can not be negative");
+            }
+            if (this.leaderEpoch != leaderEpochVal) {
```
Hmm, not sure I understand the purpose of this statement. The invariant is always `leaderEpoch == leaderEpochVal` after line 440 with or without the conditional clause.

Hangleton commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,719 @@
+                leaderEpoch = leaderEpochVal;
+            }
+            // Reset readOffset, so that it is set in next run of RLMTask
+            copiedOffsetOption = OptionalLong.empty();
```
Should the offset be reset if the leader epoch is the same as the one we already have? (it shouldn't hurt to query the highest remote offset multiple times in any case, so that is a minor observation).

Hangleton commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,719 @@
+                    // log-start-offset can be ahead of the read-offset, when:
+                    // 1) log-start-offset gets incremented via delete-records API (or)
+                    // 2) enabling the remote log for the first time
+                    long fromOffset = Math.max(copiedOffset + 1, log.logStartOffset());
```
Maybe we should log a message (warning?) if `copiedOffset + 1` < `log.logStartOffset()` in which case there may be a gap in the remote log? What if another replica (non-leader) still holds the data we are missing here?

Hangleton commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,719 @@
+                    // Copy log segments to remote storage
+                    copyLogSegmentsToRemote();
+                }
+            } catch (InterruptedException ex) {
```
One potential problem with the current SPI is that the checked `InterruptedException` cannot be propagated from plugin implementations and intercepted here.

jlprat commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -0,0 +1,719 @@
+            return threadPool;
+        }
+
+        public ScheduledFuture scheduleWithFixedDelay(Runnable runnable, long initialDelay, long delay, TimeUnit timeUnit) {
```
I know I'm late to the party, but I can see that `ScheduledFuture` is parametrized, but used "raw" in here.
I think this should be like this instead:
> public ScheduledFuture<?> scheduleWithFixedDelay(Runnable runnable, long initialDelay, long delay, TimeUnit timeUnit) {

I'll submit a fix for this
